sda-download/api/api.md

```diff
@@ -0,0 +1,161 @@
+
+### Data out API
+#### Datasets
+The `/metadata/datasets` endpoint is used to display the list of datasets the given token is authorised to access, that are present in the archive.
+##### Request
+```
+GET /metadata/datasets
+```
+##### Response
+```
+[
+    "dataset_1",
+    "dataset_2"
+]
+```
+#### Files
+##### Request
+Files contained by a dataset are listed using the `datasetName` from `/metadata/datasets`.
+```
+GET /metadata/datasets/{datasetName}/files
+```
+**Scheme Parameter**
+The `?scheme=` query parameter is optional. When a dataset contains a scheme, it may sometimes be problematic with reverse proxies.
+The scheme can be split from the dataset name, and supplied in a query parameter.
+```
+dataset := strings.Split("https://doi.org/abc/123", "://")
+len(dataset) // 2 -> scheme can be used
+dataset[0] // "https"
+dataset[1] // "doi.org/abc/123
+
+dataset := strings.Split("EGAD1000", "://")
+len(dataset) // 1 -> no scheme
+dataset[0] // "EGAD1000"
+```
+```
+GET /metadata/datasets/{datasetName}/files?scheme=https
+```
+##### Response
+```
+[
+    {
+        "fileId": "urn:file:1",
+        "datasetId": "dataset_1",
+        "displayFileName": "file_1.txt.c4gh",
+        "fileName": "hash",
+        "fileSize": 60,
+        "decryptedFileSize": 32,
+        "decryptedFileChecksum": "hash",
+        "decryptedFileChecksumType": "SHA256",
+        "fileStatus": "READY"
+    },
+    {
+        "fileId": "urn:file:2",
+        "datasetId": "dataset_1",
+        "displayFileName": "file_2.txt.c4gh",
+        "fileName": "hash",
+        "fileSize": 60,
+        "decryptedFileSize": 32,
+        "decryptedFileChecksum": "hash",
+        "decryptedFileChecksumType": "SHA256",
+        "fileStatus": "READY"
+    },
+]
+```
+#### File Data
+File data is downloaded using the `fileId` from `/metadata/datasets/{datasetName}/files`.
+##### Request
+```
+GET /files/{fileId}
+```
+##### Response
+Response is given as byte stream `application/octet-stream`
+```
+hello
+```
+##### Optional Query Parameters
+Parts of a file can be requested with specific byte ranges using `startCoordinate` and `endCoordinate` query parameters, e.g.:
+```
+?startCoordinate=0&endCoordinate=100
+```
+
+### S3 requests, for htsget
+
+**Parameters**:
+
+*Partial file retrieval*: 
+- `startCoordinate` Start byte position in the file. If the request is for an encrypted file, the position will be adjusted to align with the nearest data block boundary.
+- `endCoordinate` End byte position in the file. If the request is for an encrypted file, the position will be adjusted to align with the nearest data block boundary.
+
+Headers:
+
+- `Authorization: Bearer <token>` 
+- `Range: bytes=<start>-<end>`  exact positions. Overrides parameter coordinates.
+- `Client-public-key: <key>` used for re-encrypting the header of the file before sending it.
+- `Server-public-key: <key>` used in communication with htsget, for re-encrypting the header of the file.
```

Comment:
nitpicking 
- can't see to seem to find how are these handled
- do these come in the same request that they need to be separate ?
- there was a diagram how these requests looked like https://github.com/neicnordic/sda-download/blob/fbc45ed5885524032c5b0e0dac859c17a7f1af85/biohackathon/paper.md#htsget a new one how this would look like would help

Reply:
reason I am asking is that:
- i think it can be the same header name if those come in separate requests
- if it is htsget server that includes those headers the request i think we can have that header be an array
- (opinion) in general maybe it should be an array so that we allow multiple keys

Reply:
Here is the [new diagram](https://raw.githubusercontent.com/umccr/htsget-rs/crypt4gh/docs/crypt4gh/htsget-rs-crypt4gh.drawio.png), for context [see  this](https://github.com/umccr/htsget-rs/blob/crypt4gh/docs/crypt4gh/ARCHITECTURE.md#the-protocol).

Reply:
We do not use these headers yet, but the plan is to do so as soon as we have the re-encryption server connected (#698).

Reply:
then maybe we can include them with PR for that issue.

Reply:
talked with @dbampalikis and understood why these headers are like this, but still splitting them from this PR makes more sense

Reply:
Headers are removed in 8cece7d